### ISSUE TITLE
Installing a package already installed as dependency gives error

### ISSUE THREAD

**Issue by diwakar-s-maurya** [helm/helm#3163]
Situation:
- app1: standalone
- app2: depends on app1
- app3: depends on app1

I setup a helm repo for installation. If i install `app2` first, its fine. But now i cannot install `app3` or `app1` because it gives me error saying that the resources of `app1` already exist.

Question:
Should helm not detect that the `app1` is a common dependency (maybe by comparing url and version of app1 from the requirements.yaml of `app2` and `app3`) and since it has been installed already, so should be skipped while installing `app3`?

I have created an example too for easy testing.
```bash
helm repo add temp https://raw.githubusercontent.com/diwakar-s-maurya/helm-issue-test/master
helm repo update
helm install temp/app2
helm install temp/app3 # gives Error: release looming-maltese failed: configmaps "cm1" already exists
```

**Comment by benkeil** [comment]
We have the same problem. We deploy our whole application stack with helm. Our application use dependencies that also can deployed alone. So we have the following dependency hierarchy:

```
API
|--ESB-Service-1
|  |--Services-Namespace-Config
|  |--Basic-Service-1
|  |  |--MySql
|  |  `--Services-Namespace-Config
|  `--Basic-Service-2
|     |--MySql
|     `--Services-Namespace-Config
|--Basic-Service-3
|  |--MySql
|  `--Services-Namespace-Config
`--API-Namespace-Config
```

The `Services-Namespace-Config` for example includes a configmap that provides properties for centralized logging and can be used by all services in the services namespace.

In order to deploy `Basic-Service-1` in a separate namespace for integration testing, the chart needs the complete config to run standalone. 
If we want to deploy the `API` or `ESB-Service-1` we have the `Services-Namespace-Config` dependency multiple times and get an error because it will create 4 times the same configmap.

We don''t want to skip dependencies (we need the 3 MySQL dependencies with different configs) but we want to skip files (with same name and type) that are already installed by another package.

**Comment by phemmer** [comment]
Just throwing in my feedback and experience on issues like this.
There are cases where a dependency should be duplicated, and cases where it shouldn't. If you have a sub-chart which creates a namespace, that sub-chart could be included twice if it creates 2 different namespaces (uses a variable in the namespace name). But if both inclusions create the same namespace then obviously it should only try to do so once. And if the sub-chart is included twice and would create the same namespace, but with different content, then obviously there's a conflict as one is going to overwrite the other.

The way I've seen this type of problem handled in other projects is to define the values which operate as keys. Meaning if you have a chart which creates a namespace, and you have a key called `name`, then multiple inclusions of the chart would be allowed if the `name` value is different. If the `name` is the same, multiple inclusions would be allowed, but subsequent inclusions would be a no-op. And if the subsequent inclusions have different non-key values (meaning one inclusion would implement a resource differently from another), then an error would be thrown.

**Comment by benkeil** [comment]
I would say duplicates are never allowed. A chart is unique when its name, namespace and kind is equal. If you want for example 2 MySQL charts, then set for both another name.

**Comment by dpepper** [comment]
If you have the following environment:
    - Chart A listed as a dependency for chart B 
    - Chart A already installed using helm
    - Installing Chart B in the same namespace that chart A is already installed
I think in this setup it makes sense that helm will not try to install that dependency. Does helm offer any way to make that flow work?

**Comment by Adiqq** [comment]
Another example, chart A and chart B require service account with specific permissions to work and there's no point in duplicating Service Account, Role and RoleBinding for each new chart.
Currently it can't be done with helm due to `Error: release somerelease failed: serviceaccounts "someaccount" already exists`
In case of dependencies, we should be able to force allowing duplicates in e.g. requirements.yaml, otherwise it's extremely fragile and unreliable without any workaround.

**Comment by bacongobbler** [comment]
duplicate of #2759.
